Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicHandler.cpp

```diff
@@ -1576,115 +1586,221 @@ namespace AzToolsFramework
                 return AZ::Failure(AZStd::string("Input entity should be its owning Instance's container entity."));
             }
 
-            AZ_PROFILE_FUNCTION(AzToolsFramework);
+            // Detaching a prefab follows a few steps here.
+            // 1.  Detach the selected container entity from whichever instance owns it (so its parent instance)
+            // 2.  remove the prefab component from the container entity to turn it into a normal entity.
+            // 3.  (If keeping the container) reattach the container entity to the parent as a normal entity
+            // 4.  For each loose entity in the container prefab, reattach them to the parent instead as normal entities.
+            // 5.  For each prefab instance in the container entity, relink them to the parent instead of the container.
+            // 6.  Reorder entities to maintain the proper order
+            // 7.  (If not keeping the container) destroy the container entity.
 
+            AZ_PROFILE_FUNCTION(AzToolsFramework);
             {
-                AZ_PROFILE_SCOPE(AzToolsFramework, "Internal::DetachPrefab::UndoCapture");
+                ScopedUndoBatch undoBatch("Detach Prefab");  // outer undo is for the entire thing - detach AND (optional) delete
 
-                ScopedUndoBatch undoBatch("Detach Prefab");
-
-                InstanceOptionalReference getParentInstanceResult = owningInstance->get().GetParentInstance();
-                AZ_Assert(getParentInstanceResult.has_value(), "Can't get parent Instance from Instance of given container entity.");
+                {
+                    ScopedUndoBatch undoBatch("Detach Prefab - Actual Detach"); // inner undo is just for the detach part.
+                    AZ_PROFILE_SCOPE(AzToolsFramework, "Internal::DetachPrefab::UndoCapture");
```

Comment:
Two local variables named `undoBatch` in this function, which generates a warning and fails compilation.